rmc/src/service.rs

```diff
@@ -0,0 +1,420 @@
+/// and when a hash is successfully broadcast, the multisigned hash `Multisigned<H, MK>`
+/// is sent to the network.
+///
+/// A node with an instance of [`Service<H, MK, SCH>`] can initiate broadcasting a message `msg: H`
+/// by sending the [`RmcIncomingMessage::StartRmc`] variant through the network to [`Service<H, MK, SCH>`].
+/// As a result, the node signs `msg` and starts broadcasting the signed message via the network.
+/// When sufficintly many nodes initiate rmc with the same message `msg` and a node collects enough
+/// signatures to form a complete multisignature under the message, the multisigned message
+/// is sent back through the network by the instance of [`Service<H, MK, SCH>`].
+///
+/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html
+/// for a high-level description of this protocol and how it is used for fork alerts.
+pub struct Service<H, MK, SCH>
+where
+    H: Signable + Hash,
+    MK: MultiKeychain,
+    SCH: TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>>,
+{
+    incoming_messages:
+        UnboundedReceiver<RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>>,
+    outgoing_messages: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+    scheduler: SCH,
+    handler: Handler<H, MK>,
+}
+
+impl<H, MK, SCH> Service<H, MK, SCH>
+where
+    H: Signable + Hash + Eq + Clone + Debug,
+    MK: MultiKeychain,
+    SCH: TaskScheduler<RmcHash<H, MK::Signature, MK::PartialMultisignature>>,
+{
+    pub fn new(
+        incoming_messages: UnboundedReceiver<
+            RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+        >,
+        outgoing_messages: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+        scheduler: SCH,
+        handler: Handler<H, MK>,
+    ) -> Self {
+        Service {
+            incoming_messages,
+            outgoing_messages,
+            scheduler,
+            handler,
+        }
+    }
+
+    fn handle_message(
+        &mut self,
+        message: RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+    ) {
+        match message {
+            RmcIncomingMessage::StartRmc(hash) => {
+                debug!(target: LOG_TARGET, "starting rmc for {:?}", hash);
+                let unchecked = self.handler.on_start_rmc(hash);
+                self.scheduler.add_task(RmcHash::SignedHash(unchecked));
+            }
+            RmcIncomingMessage::RmcHash(RmcHash::MultisignedHash(unchecked)) => {
+                match self.handler.on_multisigned_hash(&unchecked) {
+                    Ok(Some(multisigned)) => {
+                        self.scheduler.add_task(RmcHash::MultisignedHash(unchecked));
+                        self.send_message(RmcOutgoingMessage::NewMultisigned(multisigned));
+                    }
+                    Ok(None) => {}
+                    Err(error) => {
+                        warn!(target: LOG_TARGET, "{}", error);
```

Comment:
Put some short context in the log-string so it will be easier to find source using logs.